File: signaling-server/server/golang/signaling.go
package main

import (
	"fmt"
	"golang.org/x/net/websocket"
	"io"
	"log"
	"net/http"
	"sync"
)

var clients map[*websocket.Conn]struct{} = make(map[*websocket.Conn]struct{})
var mutex sync.RWMutex

func notify(sender *websocket.Conn, message string) error {
	mutex.RLock()

	for c := range clients {
		if c != sender {
			if err := websocket.Message.Send(c, message); err != nil {
				return err
			}
		}
	}
	mutex.RUnlock()

	return nil
}

func handler(conn *websocket.Conn) {
	log.Println("-- websocket connected --")

	mutex.Lock()
	clients[conn] = struct{}{}
	mutex.Unlock()

	defer func(ws *websocket.Conn) {
		ws.Close()
		mutex.Lock()
		delete(clients, ws)
		mutex.Unlock()
	}(conn)

	var message string
	for {
		if err := websocket.Message.Receive(conn, &message); err != nil {
			if err == io.EOF {
				log.Println("- disconnected -")
			} else {
				log.Println(fmt.Sprintf("- unexpected error: %v -", err))
			}
			return
		}

		log.Println(fmt.Sprintf("- %s -", message))
		if err := notify(conn, message); err != nil {
			log.Println("notify error: ", err)
			return
		}
	}
}

func main() {
	s := websocket.Server{Handler: handler}
	http.Handle("/", s)
	err := http.ListenAndServe(":3001", nil)
	if err != nil {
		panic("ListenAndServe: " + err.Error())
	}

	log.Println("websocket server start. port=3001")
}

File: signaling-server/server/python/signaling.py
import asyncio
import websockets
import json

CLIENTS = set()


async def notify(sender, message):
    clients = [c for c in CLIENTS if c != sender]
    for c in clients:
        await c.send(message)


async def register_client(ws):
    CLIENTS.add(ws)


async def unregister_client(ws):
    CLIENTS.remove(ws)


async def handler(ws, path):
    print("-- websocket connected --")
    await register_client(ws)
    try:
        async for message in ws:
            print(f"< {message}")
            await notify(ws, message)
    finally:
        await unregister_client(ws)
        print("- disconnected -")


start_server = websockets.serve(handler, None, 3001)

asyncio.get_event_loop().run_until_complete(start_server)
asyncio.get_event_loop().run_forever()
